### PR TITLE
price: fix rounding of negative values

### LINKED ISSUE #346: price: rounding of negative prices
Similar to  #336 the rounding function of commerce seems to have a general issue with negative numbers.

> # Half Round Up (including negative numbers)
>
> 7.6 rounds up to 8
>7.5 rounds up to 8
>7.4 rounds down to 7
>-7.4 rounds up to -7
>-7.5 rounds down to -8
>-7.6 rounds down to -8

> # Half Round Down (including negative numbers)
>7.6 rounds up to 8
>7.5 rounds down to 7
>7.4 rounds down to 7
>-7.4 rounds up to -7
>-7.5 rounds up to -7
>-7.6 rounds down to -8

This leads to the following test code which fails:
```go
func TestRounding(t *testing.T) {
	price := domain.NewFromFloat(-7.4, "EUR")
	payable := price.GetPayableByRoundingMode(domain.RoundingModeHalfUp, 1)
	assert.EqualValues(t, domain.NewFromInt(-7, 1, "EUR").Amount(), payable.Amount())

	price = domain.NewFromFloat(-7.5, "EUR")
	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfUp, 1)
	assert.EqualValues(t, domain.NewFromInt(-8, 1, "EUR").Amount(), payable.Amount())

	price = domain.NewFromFloat(-7.6, "EUR")
	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfUp, 1)
	assert.EqualValues(t, domain.NewFromInt(-8, 1, "EUR").Amount(), payable.Amount())

	price = domain.NewFromFloat(-7.4, "EUR")
	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfDown, 1)
	assert.EqualValues(t, domain.NewFromInt(-7, 1, "EUR").Amount(), payable.Amount())

	price = domain.NewFromFloat(-7.5, "EUR")
	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfDown, 1)
	assert.EqualValues(t, domain.NewFromInt(-7, 1, "EUR").Amount(), payable.Amount())

	price = domain.NewFromFloat(-7.6, "EUR")
	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfDown, 1)
	assert.EqualValues(t, domain.NewFromInt(-8, 1, "EUR").Amount(), payable.Amount())
}
```

### PR DESCRIPTION
**Pull request by carstendietrich**
fixes #346

### CODE REVIEW

File: price/domain/price.go [hunk 1]
```diff
@@ -326,14 +326,11 @@ func (p Price) GetPayableByRoundingMode(mode string, precision int) Price {
...
 
 	amountTruncatedFloat, _ := new(big.Float).Mul(amountForRound, p.precisionF(precision)).Float64()
-	amountTruncatedInt := int64(amountTruncatedFloat)
-
-	amountRoundingCheckFloat, _ := new(big.Float).Mul(amountForRound, offsetToCheckRounding).Float64()
-	amountRoundingCheckInt := int64(amountRoundingCheckFloat)
-	valueAfterPrecision := (amountRoundingCheckInt - (amountTruncatedInt * 10)) * negative
+	integerPart, fractionalPart := math.Modf(amountTruncatedFloat)
+	amountTruncatedInt := int64(integerPart)
+	valueAfterPrecision := (math.Round(fractionalPart*10*100) / 100) * float64(negative)
```
Comment (tessig) [inline price/domain/price.go:332]:
```suggestion
	valueAfterPrecision := (math.Round(fractionalPart*1000) / 100) * float64(negative)
```

File: price/domain/price.go [hunk 2]
```diff
@@ -346,15 +343,15 @@ func (p Price) GetPayableByRoundingMode(mode string, precision int) Price {
 			amountTruncatedInt = amountTruncatedInt + (1 * negative)
```
Comment (tessig) [inline price/domain/price.go:342]:
```suggestion
			amountTruncatedInt = amountTruncatedInt + negative
```

File: price/domain/price.go [hunk 3]
```diff
@@ -346,15 +343,15 @@ func (p Price) GetPayableByRoundingMode(mode string, precision int) Price {
...
 		}
 	case mode == RoundingModeHalfUp:
-		if negative == 1 && valueAfterPrecision >= 5 {
+		if valueAfterPrecision >= 5 {
 			amountTruncatedInt = amountTruncatedInt + (1 * negative)
```
Comment (tessig) [inline price/domain/price.go:346]:
```suggestion
			amountTruncatedInt = amountTruncatedInt + negative
```

File: price/domain/price.go [hunk 4]
```diff
@@ -346,15 +343,15 @@ func (p Price) GetPayableByRoundingMode(mode string, precision int) Price {
...
 		}
 	case mode == RoundingModeHalfDown:
-		if negative == 1 && valueAfterPrecision > 5 {
+		if valueAfterPrecision > 5 {
 			amountTruncatedInt = amountTruncatedInt + (1 * negative)
```
Comment (tessig) [inline price/domain/price.go:350]:
```suggestion
			amountTruncatedInt = amountTruncatedInt + negative
```

File: price/domain/price.go [hunk 5]
```diff
@@ -346,15 +343,15 @@ func (p Price) GetPayableByRoundingMode(mode string, precision int) Price {
...
 		}
 	case mode == RoundingModeFloor:
-		if negative == -1 {
+		if negative == -1 && valueAfterPrecision > 0 {
 			amountTruncatedInt = amountTruncatedInt + (1 * negative)
```
Comment (tessig) [inline price/domain/price.go:354]:
```suggestion
			amountTruncatedInt = amountTruncatedInt + negative
```

File: price/domain/price.go (hunk 2 above)
Comment (tessig) [inline price/domain/price.go:342]: Let's take the opportunity to improve readability by using `switch mode`

File: price/domain/price_test.go [hunk 6]
```diff
@@ -196,30 +197,148 @@ func TestPrice_GetPayableByRoundingMode(t *testing.T) {
...
+
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("rounding %f", tt.price), func(t *testing.T) {
+			price := domain.NewFromFloat(tt.price, "EUR")
 
-	price = domain.NewFromFloat(-0.115, "EUR")
-	payable = price.GetPayableByRoundingMode(domain.RoundingModeHalfUp, 100)
-	assert.EqualValues(t, domain.NewFromInt(-11, 100, "EUR").Amount(), payable.Amount())
+			payable := price.GetPayableByRoundingMode(domain.RoundingModeHalfUp, tt.precision)
+			assert.Equal(t, domain.NewFromInt(tt.expected, tt.precision, "EUR").Amount(), payable.Amount(), tt.msg)
```
Comment (tessig) [inline price/domain/price_test.go:302]: I would prefer to generate all these messages on the fly:
```suggestion
			assert.Equal(t, domain.NewFromInt(tt.expected, tt.precision, "EUR").Amount(), payable.Amount(), fmt.Sprintf("%f should be rounded to %d", tt.price, tt.expected))
```

File: price/domain/price_test.go [hunk 7]
```diff
@@ -196,30 +197,148 @@ func TestPrice_GetPayableByRoundingMode(t *testing.T) {
...
+		{price: -1.6, precision: 1, expected: -2, msg: "-1.6 should be rounded to -2"},
+		{price: -2.5, precision: 1, expected: -2, msg: "-2.5 should be rounded to -2"},
+		{price: -5.5, precision: 1, expected: -5, msg: "-5.5 should be rounded to -5"},
+	}
+	for _, tt := range tests {
+		t.Run(fmt.Sprintf("rounding %f", tt.price), func(t *testing.T) {
+			price := domain.NewFromFloat(tt.price, "EUR")
+
+			payable := price.GetPayableByRoundingMode(domain.RoundingModeHalfDown, tt.precision)
+			assert.Equal(t, domain.NewFromInt(tt.expected, tt.precision, "EUR").Amount(), payable.Amount(), tt.msg)
```
Comment (tessig) [inline price/domain/price_test.go:339]: same here

File: price/domain/price.go [hunk 8]
```diff
@@ -312,7 +312,7 @@ func (p Price) GetPayable() Price {
 	return p.GetPayableByRoundingMode(mode, precision)
 }
 
-// GetPayableByRoundingMode - a flexible rounding method where you can pass rounding mode and precision
+// GetPayableByRoundingMode returns the price rounded you can pass the used rounding mode and precision
```
Comment (tessig) [inline price/domain/price.go:315]:
```suggestion
// GetPayableByRoundingMode returns the price rounded you can pass the used rounding mode and precision
// Example for precision 100:
```

File: price/domain/price.go [hunk 9]
```diff
@@ -312,7 +312,7 @@ func (p Price) GetPayable() Price {
...
 // 1.115 >  1.12 (RoundingModeHalfUp)  / 1.11 (RoundingModeFloor)
```
Comment (tessig) [inline price/domain/price.go:316]:
```suggestion
//	1.115 >  1.12 (RoundingModeHalfUp)  / 1.11 (RoundingModeFloor)
```

File: price/domain/price.go [hunk 10]
```diff
@@ -312,7 +312,7 @@ func (p Price) GetPayable() Price {
...
 // -1.115 > -1.11 (RoundingModeHalfUp) / -1.12 (RoundingModeFloor)
```
Comment (tessig) [inline price/domain/price.go:317]:
```suggestion
//	-1.115 > -1.11 (RoundingModeHalfUp) / -1.12 (RoundingModeFloor)
```

**Review by tessig — CHANGES_REQUESTED** [review]
Thanks for taking on this highly inconvenient topic. I just have a few suggestions to improve readability of the code

File: price/domain/price_test.go (hunk 6 above)
Comment (carstendietrich) [reply price/domain/price_test.go:302]: not possible due to test cases with precision != 1

**Review by tessig — APPROVED** [review]

**Pull request merged**